File: codex-v0.1.0/sources/com/poozh/translator/capture/ScreenCaptureController.java
package com.poozh.translator.capture;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.hardware.display.VirtualDisplay;
import android.media.Image;
import android.media.ImageReader;
import android.media.projection.MediaProjection;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.util.DisplayMetrics;
import java.nio.ByteBuffer;
import kotlin.Metadata;
import kotlin.Result;
import kotlin.ResultKt;
import kotlin.Unit;
import kotlin.jdk7.AutoCloseableKt;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.internal.Intrinsics;
import kotlin.ranges.RangesKt;

/* compiled from: ScreenCaptureController.kt */
@Metadata(d1 = {"\u0000m\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0010\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000b\n\u0000\n\u0002\b\t\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\b\n\u0002\b\u0004*\u0001\u001d\u0018\u0000 -2\u00020\u0001:\u0001-BO\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u000e\u0010\u0006\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\b0\u0007\u0012\u0012\u0010\t\u001a\u000e\u0012\u0004\u0012\u00020\u000b\u0012\u0004\u0012\u00020\f0\n\u0012\u0012\u0010\r\u001a\u000e\u0012\u0004\u0012\u00020\u000e\u0012\u0004\u0012\u00020\f0\n¢\u0006\u0004\b\u000f\u0010\u0010J\u0006\u0010\u001f\u001a\u00020\fJ\u0006\u0010 \u001a\u00020\fJ\u0006\u0010!\u001a\u00020\fJ\b\u0010\"\u001a\u00020\fH\u0002J\b\u0010#\u001a\u00020\u001bH\u0002J\u0010\u0010$\u001a\u00020\u000b2\u0006\u0010%\u001a\u00020&H\u0002J \u0010'\u001a\u00020\b2\u0006\u0010(\u001a\u00020\b2\u0006\u0010)\u001a\u00020*2\u0006\u0010+\u001a\u00020*H\u0002J\b\u0010,\u001a\u00020\fH\u0002R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004¢\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004¢\u0006\u0002\n\u0000R\u0016\u0010\u0006\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\b0\u0007X\u0082\u0004¢\u0006\u0002\n\u0000R\u001a\u0010\t\u001a\u000e\u0012\u0004\u0012\u00020\u000b\u0012\u0004\u0012\u00020\f0\nX\u0082\u0004¢\u0006\u0002\n\u0000R\u001a\u0010\r\u001a\u000e\u0012\u0004\u0012\u00020\u000e\u0012\u0004\u0012\u00020\f0\nX\u0082\u0004¢\u0006\u0002\n\u0000R\u000e\u0010\u0011\u001a\u00020\u0012X\u0082\u0004¢\u0006\u0002\n\u0000R\u0010\u0010\u0013\u001a\u0004\u0018\u00010\u0014X\u0082\u000e¢\u0006\u0002\n\u0000R\u0010\u0010\u0015\u001a\u0004\u0018\u00010\u0012X\u0082\u000e¢\u0006\u0002\n\u0000R\u0010\u0010\u0016\u001a\u0004\u0018\u00010\u0017X\u0082\u000e¢\u0006\u0002\n\u0000R\u0010\u0010\u0018\u001a\u0004\u0018\u00010\u0019X\u0082\u000e¢\u0006\u0002\n\u0000R\u000e\u0010\u001a\u001a\u00020\u001bX\u0082\u000e¢\u0006\u0002\n\u0000R\u0010\u0010\u001c\u001a\u00020\u001dX\u0082\u0004¢\u0006\u0004\n\u0002\u0010\u001e¨\u0006."}, d2 = {"Lcom/poozh/translator/capture/ScreenCaptureController;", "", "context", "Landroid/content/Context;", "mediaProjection", "Landroid/media/projection/MediaProjection;", "selectionProvider", "Lkotlin/Function0;", "Landroid/graphics/Rect;", "frameCallback", "Lkotlin/Function1;", "Landroid/graphics/Bitmap;", "", "errorCallback", "", "<init>", "(Landroid/content/Context;Landroid/media/projection/MediaProjection;Lkotlin/jvm/functions/Function0;Lkotlin/jvm/functions/Function1;Lkotlin/jvm/functions/Function1;)V", "mainHandler", "Landroid/os/Handler;", "captureThread", "Landroid/os/HandlerThread;", "captureHandler", "imageReader", "Landroid/media/ImageReader;", "virtualDisplay", "Landroid/hardware/display/VirtualDisplay;", "pendingCapture", "", "projectionCallback", "com/poozh/translator/capture/ScreenCaptureController$projectionCallback$1", "Lcom/poozh/translator/capture/ScreenCaptureController$projectionCallback$1;", "prepare", "captureOnce", "release", "ensureDisplay", "captureLatestFrame", "imageToBitmap", "image", "Landroid/media/Image;", "sanitizeRect", "rect", "maxWidth", "", "maxHeight", "releaseDisplay", "Companion", "app_debug"}, k = 1, mv = {2, 0, 0}, xi = 48)
/* loaded from: classes5.dex */
public final class ScreenCaptureController {
    private static final long INITIAL_FRAME_DELAY_MS = 120;
    private Handler captureHandler;
    private HandlerThread captureThread;
    private final Context context;
    private final Function1<String, Unit> errorCallback;
    private final Function1<Bitmap, Unit> frameCallback;
    private ImageReader imageReader;
    private final Handler mainHandler;
    private final MediaProjection mediaProjection;
    private boolean pendingCapture;
    private final ScreenCaptureController$projectionCallback$1 projectionCallback;
    private final Function0<Rect> selectionProvider;
    private VirtualDisplay virtualDisplay;

    /* JADX WARN: Multi-variable type inference failed */
    public ScreenCaptureController(Context context, MediaProjection mediaProjection, Function0<Rect> selectionProvider, Function1<? super Bitmap, Unit> frameCallback, Function1<? super String, Unit> errorCallback) {
        Intrinsics.checkNotNullParameter(context, "context");
        Intrinsics.checkNotNullParameter(mediaProjection, "mediaProjection");
        Intrinsics.checkNotNullParameter(selectionProvider, "selectionProvider");
        Intrinsics.checkNotNullParameter(frameCallback, "frameCallback");
        Intrinsics.checkNotNullParameter(errorCallback, "errorCallback");
        this.context = context;
        this.mediaProjection = mediaProjection;
        this.selectionProvider = selectionProvider;
        this.frameCallback = frameCallback;
        this.errorCallback = errorCallback;
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.projectionCallback = new ScreenCaptureController$projectionCallback$1(this);
    }

    public final void prepare() {
        ensureDisplay();
    }

    public final void captureOnce() {
        if (this.pendingCapture) {
            this.mainHandler.post(new Runnable() { // from class: com.poozh.translator.capture.ScreenCaptureController$$ExternalSyntheticLambda2
                @Override // java.lang.Runnable
                public final void run() {
                    ScreenCaptureController.captureOnce$lambda$0(ScreenCaptureController.this);
                }
            });
            return;
        }
        ensureDisplay();
        this.pendingCapture = true;
        Handler handler = this.captureHandler;
        if (handler != null) {
            handler.postDelayed(new Runnable() { // from class: com.poozh.translator.capture.ScreenCaptureController$$ExternalSyntheticLambda3
                @Override // java.lang.Runnable
                public final void run() {
                    ScreenCaptureController.captureOnce$lambda$2(ScreenCaptureController.this);
                }
            }, INITIAL_FRAME_DELAY_MS);
        }
    }

    /* JADX INFO: Access modifiers changed from: private */
    public static final void captureOnce$lambda$0(ScreenCaptureController this$0) {
        this$0.errorCallback.invoke("正在截取屏幕，请稍候");
    }

    /* JADX INFO: Access modifiers changed from: private */
    public static final void captureOnce$lambda$2(final ScreenCaptureController this$0) {
        boolean delivered = this$0.captureLatestFrame();
        this$0.pendingCapture = false;
        if (!delivered) {
            this$0.mainHandler.post(new Runnable() { // from class: com.poozh.translator.capture.ScreenCaptureController$$ExternalSyntheticLambda1
                @Override // java.lang.Runnable
                public final void run() {
                    ScreenCaptureController.captureOnce$lambda$2$lambda$1(ScreenCaptureController.this);
                }
            });
        }
    }

    /* JADX INFO: Access modifiers changed from: private */
    public static final void captureOnce$lambda$2$lambda$1(ScreenCaptureController this$0) {
        this$0.errorCallback.invoke("暂时没有可用画面，请再点一次刷新");
    }

    public final void release() {
        this.pendingCapture = false;
        releaseDisplay();
        try {
            Result.Companion companion = Result.INSTANCE;
            ScreenCaptureController screenCaptureController = this;
            screenCaptureController.mediaProjection.unregisterCallback(screenCaptureController.projectionCallback);
            Result.m17constructorimpl(Unit.INSTANCE);
        } catch (Throwable th) {
            Result.Companion companion2 = Result.INSTANCE;
            Result.m17constructorimpl(ResultKt.createFailure(th));
        }
        try {
            Result.Companion companion3 = Result.INSTANCE;
            this.mediaProjection.stop();
            Result.m17constructorimpl(Unit.INSTANCE);
        } catch (Throwable th2) {
            Result.Companion companion4 = Result.INSTANCE;
            Result.m17constructorimpl(ResultKt.createFailure(th2));
        }
    }

    private final void ensureDisplay() {
        if (this.virtualDisplay != null) {
            return;
        }
        DisplayMetrics metrics = this.context.getResources().getDisplayMetrics();
        int width = Math.max(metrics.widthPixels, 1);
        int height = Math.max(metrics.heightPixels, 1);
        int density = metrics.densityDpi;
        HandlerThread thread = new HandlerThread("screen-translator-capture");
        thread.start();
        this.captureThread = thread;
        this.captureHandler = new Handler(thread.getLooper());
        ImageReader reader = ImageReader.newInstance(width, height, 1, 2);
        Intrinsics.checkNotNullExpressionValue(reader, "newInstance(...)");
        this.imageReader = reader;
        this.mediaProjection.registerCallback(this.projectionCallback, this.mainHandler);
        this.virtualDisplay = this.mediaProjection.createVirtualDisplay("ScreenTranslator", width, height, density, 16, reader.getSurface(), null, this.captureHandler);
    }

    private final boolean captureLatestFrame() {
        Object m17constructorimpl;
        Rect rect;
        ImageReader reader = this.imageReader;
        if (reader == null) {
            return false;
        }
        try {
            Result.Companion companion = Result.INSTANCE;
            ScreenCaptureController screenCaptureController = this;
            m17constructorimpl = Result.m17constructorimpl(reader.acquireLatestImage());
        } catch (Throwable th) {
            Result.Companion companion2 = Result.INSTANCE;
            m17constructorimpl = Result.m17constructorimpl(ResultKt.createFailure(th));
        }
        if (Result.m23isFailureimpl(m17constructorimpl)) {
            m17constructorimpl = null;
        }
        Image image = (Image) m17constructorimpl;
        if (image == null) {
            return false;
        }
        Image image2 = image;
        try {
            Bitmap imageToBitmap = imageToBitmap(image2);
            Rect invoke = this.selectionProvider.invoke();
            if (invoke == null || (rect = sanitizeRect(invoke, imageToBitmap.getWidth(), imageToBitmap.getHeight())) == null) {
                rect = new Rect(0, 0, imageToBitmap.getWidth(), imageToBitmap.getHeight());
            }
            if (rect.width() >= 8 && rect.height() >= 8) {
                final Bitmap createBitmap = Bitmap.createBitmap(imageToBitmap, rect.left, rect.top, rect.width(), rect.height());
                Intrinsics.checkNotNullExpressionValue(createBitmap, "createBitmap(...)");
                imageToBitmap.recycle();
                this.mainHandler.post(new Runnable() { // from class: com.poozh.translator.capture.ScreenCaptureController$$ExternalSyntheticLambda0
                    @Override // java.lang.Runnable
                    public final void run() {
                        ScreenCaptureController.captureLatestFrame$lambda$9$lambda$8(ScreenCaptureController.this, createBitmap);
                    }
                });
                AutoCloseableKt.closeFinally(image2, null);
                return true;
            }
            imageToBitmap.recycle();
            AutoCloseableKt.closeFinally(image2, null);
            return false;
        } finally {
        }
    }

    /* JADX INFO: Access modifiers changed from: private */
    public static final void captureLatestFrame$lambda$9$lambda$8(ScreenCaptureController this$0, Bitmap $cropped) {
        this$0.frameCallback.invoke($cropped);
    }

    private final Bitmap imageToBitmap(Image image) {
        Image.Plane plane = image.getPlanes()[0];
        ByteBuffer buffer = plane.getBuffer();
        buffer.rewind();
        int pixelStride = plane.getPixelStride();
        int rowStride = plane.getRowStride();
        int rowPadding = rowStride - (image.getWidth() * pixelStride);
        int paddedWidth = image.getWidth() + (rowPadding / pixelStride);
        Bitmap paddedBitmap = Bitmap.createBitmap(paddedWidth, image.getHeight(), Bitmap.Config.ARGB_8888);
        Intrinsics.checkNotNullExpressionValue(paddedBitmap, "createBitmap(...)");
        paddedBitmap.copyPixelsFromBuffer(buffer);
        Bitmap bitmap = Bitmap.createBitmap(paddedBitmap, 0, 0, image.getWidth(), image.getHeight());
        Intrinsics.checkNotNullExpressionValue(bitmap, "createBitmap(...)");
        paddedBitmap.recycle();
        return bitmap;
    }

    private final Rect sanitizeRect(Rect rect, int maxWidth, int maxHeight) {
        int left = RangesKt.coerceIn(rect.left, 0, maxWidth - 1);
        int top = RangesKt.coerceIn(rect.top, 0, maxHeight - 1);
        int right = RangesKt.coerceIn(rect.right, left + 1, maxWidth);
        int bottom = RangesKt.coerceIn(rect.bottom, top + 1, maxHeight);
        return new Rect(left, top, right, bottom);
    }

    /* JADX INFO: Access modifiers changed from: private */
    public final void releaseDisplay() {
        Unit unit;
        Unit unit2;
        try {
            Result.Companion companion = Result.INSTANCE;
            VirtualDisplay virtualDisplay = this.virtualDisplay;
            if (virtualDisplay != null) {
                virtualDisplay.release();
                unit2 = Unit.INSTANCE;
            } else {
                unit2 = null;
            }
            Result.m17constructorimpl(unit2);
        } catch (Throwable th) {
            Result.Companion companion2 = Result.INSTANCE;
            Result.m17constructorimpl(ResultKt.createFailure(th));
        }
        this.virtualDisplay = null;
        try {
            Result.Companion companion3 = Result.INSTANCE;
            ImageReader imageReader = this.imageReader;
            if (imageReader != null) {
                imageReader.close();
                unit = Unit.INSTANCE;
            } else {
                unit = null;
            }
            Result.m17constructorimpl(unit);
        } catch (Throwable th2) {
            Result.Companion companion4 = Result.INSTANCE;
            Result.m17constructorimpl(ResultKt.createFailure(th2));
        }
        this.imageReader = null;
        Handler handler = this.captureHandler;
        if (handler != null) {
            handler.removeCallbacksAndMessages(null);
        }
        HandlerThread handlerThread = this.captureThread;
        if (handlerThread != null) {
            handlerThread.quitSafely();
        }
        this.captureHandler = null;
        this.captureThread = null;
    }
}

File: app/src/main/java/com/poozh/translator/capture/ScreenCaptureController.kt
package com.poozh.translator.capture

import android.content.Context
import android.graphics.Bitmap
import android.graphics.PixelFormat
import android.graphics.Rect
import android.hardware.display.DisplayManager
import android.hardware.display.VirtualDisplay
import android.media.Image
import android.media.ImageReader
import android.media.projection.MediaProjection
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.view.WindowManager
import kotlin.math.max
import kotlin.math.min

class ScreenCaptureController(
    private val context: Context,
    private val mediaProjection: MediaProjection,
    private val selectionProvider: () -> Rect?,
    private val frameCallback: (Bitmap) -> Unit,
    private val errorCallback: (String) -> Unit,
    /** Fired on the main thread when the system revokes the MediaProjection
     *  (token expired, user revoked, single-use on Android 14+). The service
     *  uses this to null out its [mediaProjection] reference so the next refresh
     *  re-requests permission instead of looping on a dead projection. */
    private val onProjectionStopped: () -> Unit = {}
) {
    private val mainHandler = Handler(Looper.getMainLooper())
    private var captureThread: HandlerThread? = null
    private var captureHandler: Handler? = null
    private var imageReader: ImageReader? = null
    private var virtualDisplay: VirtualDisplay? = null
    private var pendingCapture = false

    private val projectionCallback = object : MediaProjection.Callback() {
        override fun onStop() {
            mainHandler.post {
                pendingCapture = false
                releaseDisplay()
                // Tell the service the projection is dead so it clears its
                // mediaProjection field and re-requests permission on next use.
                onProjectionStopped()
                errorCallback("屏幕捕获授权已失效，正在重新请求…")
            }
        }
    }

    fun prepare() {
        ensureDisplay()
    }

    fun captureOnce() {
        if (pendingCapture) {
            mainHandler.post { errorCallback("正在截取屏幕，请稍候") }
            return
        }
        ensureDisplay()
        drainPendingImages()
        pendingCapture = true
        captureHandler?.postDelayed({
            val delivered = captureLatestFrame()
            pendingCapture = false
            if (!delivered) {
                mainHandler.post { errorCallback("暂时没有可用画面，请再点一次刷新") }
            }
        }, INITIAL_FRAME_DELAY_MS)
    }

    fun release() {
        pendingCapture = false
        releaseDisplay()
        runCatching { mediaProjection.unregisterCallback(projectionCallback) }
        runCatching { mediaProjection.stop() }
    }

    private fun ensureDisplay() {
        if (virtualDisplay != null) return

        // Use the REAL full-screen size (including system bars), not the
        // app-window size from displayMetrics. The selection overlay is a
        // MATCH_PARENT window covering the entire physical screen, so its
        // view-local coordinates run from 0..realWidth/realHeight. If we size
        // the ImageReader from displayMetrics.heightPixels (which excludes the
        // status/navigation bars on most devices), the captured bitmap is
        // shorter than the selection coordinate space, and the crop ends up
        // misaligned (the recognised region slides up vs. the drawn box).
        val (width, height) = realScreenSize()
        val density = context.resources.displayMetrics.densityDpi

        val thread = HandlerThread("screen-translator-capture").also { it.start() }
        captureThread = thread
        captureHandler = Handler(thread.looper)
        val reader = ImageReader.newInstance(width, height, PixelFormat.RGBA_8888, 2)
        imageReader = reader

        mediaProjection.registerCallback(projectionCallback, mainHandler)
        // createVirtualDisplay throws SecurityException / IllegalStateException on
        // Android 14+ if the projection isn't active or the FGS type isn't allowed.
        // Wrap it so a failure cleans up half-built resources and reports a clear
        // error instead of crashing.
        virtualDisplay = try {
            mediaProjection.createVirtualDisplay(
                "ScreenTranslator",
                width,
                height,
                density,
                DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
                reader.surface,
                null,
                captureHandler
            )
        } catch (e: Throwable) {
            runCatching { mediaProjection.unregisterCallback(projectionCallback) }
            releaseDisplay()
            // A SecurityException here almost always means the projection token
            // is dead — fire onProjectionStopped so the service clears its
            // reference and re-requests permission instead of looping.
            mainHandler.post {
                onProjectionStopped()
                errorCallback("屏幕捕获授权已失效，正在重新请求…")
            }
            null
        }
    }

    private fun captureLatestFrame(): Boolean {
        val reader = imageReader ?: run {
            android.util.Log.w(TAG, "captureLatestFrame: imageReader is null")
            return false
        }
        val image = runCatching { reader.acquireLatestImage() }.getOrNull() ?: run {
            android.util.Log.w(TAG, "captureLatestFrame: no frame available yet")
            return false
        }
        image.use { current ->
            val bitmap = imageToBitmap(current)
            val rawSelection = selectionProvider()
            val cropRect = rawSelection?.let { sanitizeRect(it, bitmap.width, bitmap.height) }
                ?: Rect(0, 0, bitmap.width, bitmap.height)
            android.util.Log.d(TAG, "captureLatestFrame: bitmap=${bitmap.width}x${bitmap.height}, selection=$rawSelection, cropRect=$cropRect")
            if (cropRect.width() < 8 || cropRect.height() < 8) {
                android.util.Log.w(TAG, "captureLatestFrame: cropRect too small, dropping")
                bitmap.recycle()
                return false
            }
            val cropped = Bitmap.createBitmap(bitmap, cropRect.left, cropRect.top, cropRect.width(), cropRect.height())
            bitmap.recycle()
            android.util.Log.d(TAG, "captureLatestFrame: cropped=${cropped.width}x${cropped.height}, delivering")
            mainHandler.post { frameCallback(cropped) }
            return true
        }
    }

    private fun drainPendingImages() {
        val reader = imageReader ?: return
        while (true) {
            val staleImage = runCatching { reader.acquireLatestImage() }.getOrNull() ?: return
            staleImage.close()
        }
    }

    private fun imageToBitmap(image: Image): Bitmap {
        val plane = image.planes[0]
        val buffer = plane.buffer
        buffer.rewind()
        val pixelStride = plane.pixelStride
        val rowStride = plane.rowStride
        val rowPadding = rowStride - pixelStride * image.width
        val paddedWidth = image.width + rowPadding / pixelStride
        val paddedBitmap = Bitmap.createBitmap(paddedWidth, image.height, Bitmap.Config.ARGB_8888)
        paddedBitmap.copyPixelsFromBuffer(buffer)
        val bitmap = Bitmap.createBitmap(paddedBitmap, 0, 0, image.width, image.height)
        paddedBitmap.recycle()
        return bitmap
    }

    private fun sanitizeRect(rect: Rect, maxWidth: Int, maxHeight: Int): Rect {
        val left = rect.left.coerceIn(0, maxWidth - 1)
        val top = rect.top.coerceIn(0, maxHeight - 1)
        val right = rect.right.coerceIn(left + 1, maxWidth)
        val bottom = rect.bottom.coerceIn(top + 1, maxHeight)
        return Rect(left, top, right, bottom)
    }

    /**
     * The real, full physical screen size including system bars. Matches the
     * coordinate space of a top-left-aligned MATCH_PARENT overlay window (which
     * is what the selection view is), so selection coordinates map 1:1 onto the
     * captured bitmap.
     */
    private fun realScreenSize(): Pair<Int, Int> {
        val wm = context.getSystemService(Context.WINDOW_SERVICE) as? WindowManager
        if (wm != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val bounds = wm.maximumWindowMetrics.bounds
            val w = max(bounds.width(), 1)
            val h = max(bounds.height(), 1)
            if (w > 0 && h > 0) return w to h
        }
        // Fallback for older APIs: real metrics (includes system bars).
        val metrics = android.util.DisplayMetrics()
        @Suppress("DEPRECATION")
        (context.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.getRealMetrics(metrics)
        return max(metrics.widthPixels, 1) to max(metrics.heightPixels, 1)
    }

    private fun releaseDisplay() {
        runCatching { virtualDisplay?.release() }
        virtualDisplay = null
        runCatching { imageReader?.close() }
        imageReader = null
        captureHandler?.removeCallbacksAndMessages(null)
        captureThread?.quitSafely()
        captureHandler = null
        captureThread = null
    }

    companion object {
        private const val TAG = "ScreenCaptureCtrl"
        private const val INITIAL_FRAME_DELAY_MS = 240L
        /** How long to wait after flushing the buffer before grabbing the frame
         *  we actually OCR — lets the compositor settle after a transition. */
        private const val SETTLE_DELAY_MS = 150L
    }
}
